File: files/100-199/182/CH-182.py
import pandas as pd

path = "CH-182 Indexing Blank cells.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=1, nrows=12, dtype=str)
test = pd.read_excel(path, usecols="F:H", skiprows=1, nrows=12, dtype=str).rename(columns=lambda x: x.split('.')[0])

flat_values = input.values.flatten(order='C')

counter = 1
for i in range(len(flat_values)):
    if pd.isna(flat_values[i]):
        flat_values[i] = f'B{counter}'
        counter += 1

result = pd.DataFrame(flat_values.reshape(input.shape), columns = input.columns)

print(result.equals(test)) # True

File: files/100-199/182/CH-182.R
library(tidyverse)
library(readxl)

path = "files/CH-182 Indexing Blank cells.xlsx"
input = read_excel(path, range = "B2:D14", col_types = "text")
test  = read_excel(path, range = "F2:H14")

result = input %>%
  mutate(rn = row_number()) %>%
  pivot_longer(cols = -rn, names_to = "col", values_to = "value") %>%
  arrange(rn, col) %>%
  mutate(value = ifelse(is.na(value), paste0("B", cumsum(is.na(value))), value)) %>%
  pivot_wider(names_from = col, values_from = value) %>%
  select(-rn)

all.equal(result, test)
#> [1] TRUE